controllers/containerizedworkload_controller_helper.go

```diff
@@ -34,87 +29,58 @@ const (
-			},
-		},
+
+	resources, err := cwh.Translator(ctx, workload)
+	if err != nil {
+		return nil, err
+	}
+	deploy, ok := resources[0].(*appsv1.Deployment)
+	if !ok {
+		return nil, fmt.Errorf("internal error, deployment is not rendered correctly")
 	}
+	// TODO: fix bug in the lib
+	r.Log.Info("translate a deployment", "deploy", deploy.Spec.Template.Spec.Containers[0])
+	deploy.Namespace = workload.Namespace
 
-	// always set the controller reference so that we can watch this deployment
-	if err := ctrl.SetControllerReference(workload, &depl, r.Scheme); err != nil {
+	for i := 0; i < len(deploy.Spec.Template.Spec.Containers); i++ {
+		for j := 0; j < len(deploy.Spec.Template.Spec.Containers[i].Ports); j++ {
+			if len(deploy.Spec.Template.Spec.Containers[i].Ports[j].Protocol) == 0 {
+				deploy.Spec.Template.Spec.Containers[i].Ports[j].Protocol = corev1.ProtocolTCP
+			}
+		}
+	}
+	r.Log.Info(" rendered a deployment", "deploy", deploy.Spec.Template.Spec.Containers[0])
+	// always set the controller reference so that we can watch this deployment and it will be deleted automatically
+	if err := ctrl.SetControllerReference(workload, deploy, r.Scheme); err != nil {
 		return nil, err
 	}
 
-	return &depl, nil
+	return deploy, nil
 }
 
 // create a service for the deployment
-func (r *ContainerizedWorkloadReconciler) renderService(_ context.Context, deploy *appsv1.Deployment,
-	workload *oamv1alpha2.ContainerizedWorkload) (*corev1.Service, error) {
-	// create a default in case there is no container port
-	port := &corev1.ContainerPort{
-		Name:          "OAM default",
-		Protocol:      "TCP",
-		ContainerPort: 8080,
-	}
-	for _, c := range deploy.Spec.Template.Spec.Containers {
-		if len(c.Ports) != 0 {
-			port = &c.Ports[0]
-		}
+func (r *ContainerizedWorkloadReconciler) renderService(ctx context.Context,
+	workload *oamv1alpha2.ContainerizedWorkload, deploy *appsv1.Deployment) (*corev1.Service, error) {
+	// create a service for the workload
+	resources, err := wh.ServiceInjector(ctx, workload, []resource.Object{deploy})
+	if err != nil {
+		return nil, err
 	}
-
-	svc := corev1.Service{
-		TypeMeta: metav1.TypeMeta{
-			APIVersion: corev1.SchemeGroupVersion.String(),
-			Kind:       KindService,
-		},
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      deploy.Name + "-service",
-			Namespace: deploy.Namespace,
-		},
-		Spec: corev1.ServiceSpec{
-			Ports: []corev1.ServicePort{
-				{
-					Name:       strings.ToLower(string(port.Protocol)),
-					Port:       8080,
-					Protocol:   port.Protocol,
-					TargetPort: intstr.FromInt(int(port.ContainerPort)),
-				},
-			},
-			Selector: map[string]string{OAMResourceNameLabel: deploy.Name},
-			Type:     corev1.ServiceTypeClusterIP,
-		},
+	service, ok := resources[1].(*corev1.Service)
+	if !ok {
+		return nil, fmt.Errorf("internal error, service is not rendered correctly")
 	}
+	// TODO: fix it in the lib
+	service.Namespace = workload.Namespace
+	service.Spec.Ports[0].Protocol = corev1.ProtocolTCP
```

Comment:
Kubernetes will default to TCP here :)

Reply:
This is where an e2e test is critical as any UT tests use a mocked client won't catch this.  It is true in theory/doc but I spent a good amount of time deciphering the very helpful K8s error messages to figure out somehow it's not the case. I wouldn't put this ugly code here otherwise.